### PR TITLE
Brain MRI Super-Resolutionion

### PR DESCRIPTION
**Pull request by a12a12a12a12**
Brain MRI Super-Resolution using ESPCN
Zijun Zhu
 s4627546

Efficient Sub-Pixel Convolutional Neural Network (ESPCN) is designed to upscale low-resolution images to high-resolution counterparts. In this implementation, we aim to solve the problem of enhancing the resolution of Brain MRI scans. Given a down-sampled MRI image, our model attempts to produce a clearer image with a resolution upscale factor of 4.

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Overall a good attempt. Does well to create a custom dataloader to deal with AD and NC folders. Training looks good too, just unsure about the mixed precision (nothing mentioned in README about it). "predict.py" doesn't normalize before inference, but then code denormalizes, results in some strange images. Image quality better than low rez, but I suspect the lack of normalization in "predict.py" and mixed precision might be negatively impact performance.

**[Recognition Problem]**
"modules.py":

- Re-implementation of the Keras example using PyTorch. All seems ok. 

"dataset.py":

- Nice custom dataset to deal with AD and NC folder structure.
- Performs resizing correctly.
- Has some visualization to check.

"train.py":

- Uses train and validation sets.
- Uses OneCycleLR and GradScaler during training.
- Overall seems ok, just using mixed precision operations?
- Plots training and validation losses at the end.
- Saves the model with the lowest validation loss. Nice.

"predict.py":

- Has code to normalize and denormalizes to ensure images are still 0->1, however, only denormalize is called, resluting in changed contrast to images.
- Provides some figures, however, no metrics on test set are demonstrated.
- Otherwise ok.

"README.MD":

- README is in the wrong folder.
- Provides a brief description of the network and problem approach.
- Provides the file structure (no changes to the dataset).
- Provides some input/output images. Not as sharp as other attempts and has some artefacts. Perhaps a result of the mixed precision training?
- Provides training and validation losses. Looks ok.
- Provides dependencies.

**[Commit Log]**
Commit log looks ok.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Solves problem, some minor normalisation issues
Driver Script present
File structure present, but READMEs not placed correctly -2
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
Feedback required, revert changes to repo README and move current project README -2
Request Description minimal -1

**Comment by shakes76** [comment]
Feedback marks +2 can be earned by fixing the READMEs as above.

**Comment by a12a12a12a12** [comment]
> Feedback marks +2 can be earned by fixing the READMEs as above.

**Comment by a12a12a12a12** [comment]
> Feedback marks +2 can be earned by fixing the READMEs as above.

Have fixed it, thanks a lot!

**Comment by wangzhaomxy** [comment]
Feedback attempt and feedback marks granted +2.
